### PR TITLE
Ethereum/FinalChain code refactoring, Revisited Ethereum json-rpc impl., and more

### PR DESCRIPTION
**Pull request by 02p01r**
1) Ethereum/FinalChain code refactoring.
- Removed all taraxa-aleth high-level ethereum code, rewrote from scratch according to our needs, making it more compact and efficient.
- Redesigned FinalChain database schema, in particular we don't store full executed transactions twice now, and have better performance on querying executed transaction hashes. Also removed some redundant data that was stored previously (log block blooms) 
- Merged FinalChain and Executor, removed Executor.
- Put all the FinalChain-related code to one package, eliminated all it's dependencies except the DB 
2) Implemented eth json-rpc subscription API and revisited other json-rpc implementations according to the specs https://ethereum.org/en/developers/docs/apis/json-rpc/. Refactored json-rpc code.
3) Added python testing subproject, tested Ethrereum json-rpc: https://github.com/Taraxa-project/taraxa-node/pull/861.
4) Reverted synchronous db commit in taraxa-evm, now FinalChain recovery code considers taraxa-evm db being arbitrarily far behind FinalChain db.
5) Refactored DbStorage.
- Changed the batch type from `shared_ptr<Batch>` to just `Batch`
- Added some Slice converters
- Unified methods for inserting (with or without batch), reading, and deleting arbitrarily-typed keys and values
6) We now have only one Transaction class in the code base - taraxa::Transaction
7) Misc improvements all over the code.
- Added a publish/subscribe framework (`src/util/event.hpp`), used it in ethereum json-rpc subscription APIs 
- Extended the RLP framework (`src/util/encoding_rlp.hpp`): added more supported types and created a set of macros that allow to describe RLP serialization for arbitrary structures.
- In Transaction: added an ability to provide known precomputed hash to the constructor; added an indicator that a transaction was default-constructed. Used the improved RLP framework in the constructor. 

Also see https://github.com/Taraxa-project/taraxa-aleth/pull/24
Also see https://github.com/Taraxa-project/taraxa-evm/pull/28

### CODE REVIEW

**Comment by MatusKysel** [comment]
It is very strange that you increased execution time of almost all tests (except `state_api_test`), that does not look good at all

**Comment by MatusKysel** [comment]
>     1. Ethereum/FinalChain code refactoring.
> 
> 
>     * Removed all taraxa-aleth high-level ethereum code, rewrote from scratch according to our needs, making it more compact and efficient.
> 
>     * Redesigned FinalChain database schema
> 
>     * Merged FinalChain and Executor, removed Executor.
> 
>     * Put all the FinalChain-related code to one package, eliminated all it's dependencies except the DB
> 
> 
>     1. Implemented eth json-rpc subscription API and revisited other json-rpc implementations according to the specs https://ethereum.org/en/developers/docs/apis/json-rpc/. Refactored json-rpc code.
> 
>     2. Added python testing subproject, tested Ethrereum json-rpc [WIP].
> 
>     3. Reverted synchronous db commit in taraxa-evm, now FinalChain recovery code considers taraxa-evm db being arbitrarily far behind FinalChain db.
> 
>     4. Refactored DbStorage.
> 
> 
>     * Changed the batch type from `shared_ptr<Batch>` to just `Batch`
> 
>     * Added some Slice converters
> 
>     * Unified methods for inserting (with or without batch), reading, and deleting arbitrarily-typed keys and values
> 
> 
>     1. Misc improvements all over the code.
> 
> 
>     * Added a publish/subscribe framework (`src/util/event.hpp`), used it in ethereum json-rpc subscription APIs
> 
>     * Extended the RLP framework (`src/util/encoding_rlp.hpp`): added more supported types and created a set of macros that allow to describe RLP serialization for arbitrary structures.
> 
>     * In Transaction: added an ability to provide known precomputed hash to the constructor; added an indicator that a transaction was default-constructed. Used the improved RLP framework in the constructor.

I think you should split this PR exactly as you describe it here, to at least 6 separate PR, because right now you are mixing lot of unrelated staff together

**Comment by JakubFornadel** [comment]
So just by looking at the "-" commits I guess you are going to squash all the commits together and then merge as you did before on your PR's. As a result current PR description will be commit message. So 1 single commit adds 5k lines, deletes 3k lines. You are changing stuff in multiple unrelated areas of the code. Now imagine you introduced some new bug, how on earth are we supposed to find out  which part of this gigantic commit did that ? This PR should have been multiple PR's + each PR might have multiple commits that could be then easily cherry-picked, reverted, tested, etc...

File: src/CMakeLists.txt
Comment (JakubFornadel) [inline src/CMakeLists.txt:55]: put list of headers back

File: src/common/types.hpp
Comment (JakubFornadel) [inline src/common/types.hpp:28]: Having using in headers is very bad practice. You just pollute namespaces thanks to this as every file that includes "types.hpp" will contain these usings, which is pretty bad. Here are some explanation related to this: https://jmmv.dev/2013/12/header-files-avoid-c-using-directives.html
https://stackoverflow.com/questions/19940081/using-keyword-in-header-implementation-in-c

File: src/final_chain/final_chain.hpp
Comment (JakubFornadel) [inline src/final_chain/final_chain.hpp:19]: again, have these "usings" in cpp files.

File: src/network/rpc/EthClient.h
Comment (JakubFornadel) [inline src/network/rpc/EthClient.h:11]: we are using 2 space padding in whole codebase

File: src/network/rpc/NetClient.h
Comment (JakubFornadel) [inline src/network/rpc/NetClient.h:18]: we are using 2 space padding in whole codebase

File: src/network/rpc/TestClient.h
Comment (JakubFornadel) [inline src/network/rpc/TestClient.h:7]: I guess these guards were generated automatically ? As we are using pragmas but this seems like generation-related issue...

File: tests/pbft_manager_test.cpp
Comment (JakubFornadel) [inline tests/pbft_manager_test.cpp:67]: I believer we are using camel case for function names

File: tests/CMakeLists.txt
Comment (JakubFornadel) [inline tests/CMakeLists.txt:9]: is this supposed to be commented out ? How is the new test run then ?

File: src/node/full_node.hpp
Comment (JakubFornadel) [inline src/node/full_node.hpp:98]: I was looking at this before. Can you please explain me what is this even good for ?

File: src/node/full_node.cpp
Comment (JakubFornadel) [inline src/node/full_node.cpp:244]: TODO what ?

File: src/network/rpc/eth/impl/Watches.hpp
Comment (JakubFornadel) [inline src/network/rpc/eth/impl/Watches.hpp:21]: why do you use struct instead of class for such relatively complex data stucture ?

File: src/network/rpc/CMakeLists.txt
Comment (JakubFornadel) [inline src/network/rpc/CMakeLists.txt:14]: well gen_rpc_stubs is definitely more clear in case this custom target is still supposed to generate rpc stubs :)

File: src/final_chain/final_chain.cpp
Comment (JakubFornadel) [inline src/final_chain/final_chain.cpp:14]: class...

File: src/final_chain/final_chain.cpp
Comment (JakubFornadel) [inline src/final_chain/final_chain.cpp:17]: please fix this. Using "new" is just huge NO

File: src/final_chain/final_chain.cpp
Comment (JakubFornadel) [inline src/final_chain/final_chain.cpp:155]: do not use "new", use make_shared/unique instead

File: src/final_chain/final_chain.cpp
Comment (JakubFornadel) [inline src/final_chain/final_chain.cpp:437]: do not use "new", use make_shared/unique instead

**Comment by JakubFornadel** [comment]
Seems there might be actually a lot of good stuff here, just the way this PR is done degrades the overall quality a lot... I dont even know where and how to do some logical review here as there is so many lines and so much stuff mixed here... I tend to agree with Matus that this should have been split into multiple different PR's with proper commits descriptions etc...

File: src/common/types.hpp
Comment (02p01r) [reply src/common/types.hpp:28]: It is equivalent to a bunch of typedefs that we already have in that header. Therese particular types are so ubiquitous in our code that I thought it is okay to have them globally defined in our namespace. Would you be more accepting to `using h64 = dev::h64` for example (exactly same thing)? Or you think those type names should not belong to `taraxa` namespace at all?

File: src/network/rpc/EthClient.h
Comment (02p01r) [reply src/network/rpc/EthClient.h:11]: this is a generated code

File: src/network/rpc/NetClient.h
Comment (02p01r) [reply src/network/rpc/NetClient.h:18]: this is a generated code

File: src/network/rpc/TestClient.h
Comment (02p01r) [reply src/network/rpc/TestClient.h:7]: this is a generated code

File: tests/pbft_manager_test.cpp
Comment (02p01r) [reply tests/pbft_manager_test.cpp:67]: agree that the naming convention should be consistent. we have established naming for some types of members, but not for functions. I hope this is not critical to you right now, and anyway it should be fixed at once with a sweeping change.

File: tests/CMakeLists.txt
Comment (02p01r) [reply tests/CMakeLists.txt:9]: The python tests are in progress, I believe there's a TODO above those commented lines that suggests that it should be uncommented eventually.
How this test will work: we compile a launcher executable that will itself just run the script `tests/py/run.sh` which does all the job. It seems ugly that we have to create an intermediate executable for this - I made it to stick to the pattern we have here. But I can also implement it via cmake custom target if you want.

File: src/node/full_node.hpp
Comment (02p01r) [reply src/node/full_node.hpp:98]: Long ago we had been having a devastating problem in the tests: nodes would randomly not fully terminate when a test finishes, and the next test would then fail on acquiring some unreleased OS resource like network port or rocksdb db lock.
I fixed it by refactoring the node lifecycle, and put this safeguard to make sure we don't have the problem again.
The idea of it is to make sure that all the shared pointers inside full node are released when the node is destroyed.

File: src/node/full_node.cpp
Comment (02p01r) [reply src/node/full_node.cpp:244]: I will make it more clear in the comment.
TODO - remove that line. We should not care about explicitly controlling destruction of objects. The lifecycle of objects should be as declarative as possible (kind of RAII).

File: src/network/rpc/eth/impl/Watches.hpp
Comment (02p01r) [reply src/network/rpc/eth/impl/Watches.hpp:21]: The only difference between class and struct in C++ is that class has private visibility by default and struct - public.

I use class or struct keyword only with this in mind. For instance, if I want the first definitions in a class/struct be private - I will use `class`. Otherwise - `struct`

File: src/network/rpc/CMakeLists.txt
Comment (02p01r) [reply src/network/rpc/CMakeLists.txt:14]: I will rename it to gen_rpc_stubs

File: src/final_chain/final_chain.cpp
Comment (02p01r) [reply src/final_chain/final_chain.cpp:14]: https://github.com/Taraxa-project/taraxa-node/pull/843#discussion_r648265919

File: src/final_chain/final_chain.cpp
Comment (02p01r) [reply src/final_chain/final_chain.cpp:17]: ok

File: src/final_chain/final_chain.cpp
Comment (02p01r) [reply src/final_chain/final_chain.cpp:155]: Here I used `new` just because I wanted to use aggregate initialization. make_shared doesn't allow to do it unfortunately. That means I need to define a normal constructor in FinalizationResult, which I wanted to avoid. You may have noticed that I prefer aggregate structs for simple data types. 

I will find a way to avoid `new` there

File: src/final_chain/final_chain.cpp
Comment (02p01r) [reply src/final_chain/final_chain.cpp:437]: ok about make_unique.

>What is this whole NewFinalChain function even good for ?

It is the only publicly exposed thing in this .cpp file. The idea of this pattern is to completely hide the implementation. The header defines an interface + a factory (this function). And the only job .cpp file does - provide an interface implementation by defining the factory.

**Comment by 02p01r** [comment]
> Seems there might be actually a lot of good stuff here, just the way this PR is done degrades the overall quality a lot... I dont even know where and how to do some logical review here as there is so many lines and so much stuff mixed here... I tend to agree with Matus that this should have been split into multiple different PR's with proper commits descriptions etc...

As discussed offline, I will simulate a history of well-defined and build-able commits based on these changes

File: src/CMakeLists.txt
Comment (02p01r) [reply src/CMakeLists.txt:55]: Ok, but is there any particular reason for this? As far as I know there's no effect from listing headers as compilation units

**Comment by 02p01r** [comment]
> So just by looking at the "-" commits I guess you are going to squash all the commits together and then merge as you did before on your PR's. As a result current PR description will be commit message. So 1 single commit adds 5k lines, deletes 3k lines. You are changing stuff in multiple unrelated areas of the code. Now imagine you introduced some new bug, how are we supposed to find out which part of this gigantic commit did that ? This PR should have been multiple PR's + each PR might have multiple commits that could be then easily cherry-picked, reverted, tested, etc...

Right, I wanted to turn those "-" in one big description using squash.
The rest of my answer to this overlaps with this thread: https://github.com/Taraxa-project/taraxa-node/pull/843#issuecomment-857667530

File: src/storage/db_storage.hpp
Comment (MatusKysel) [inline src/storage/db_storage.hpp:73]: why no suffix *_ here?

File: src/util/encoding_rlp.hpp
Comment (MatusKysel) [inline src/util/encoding_rlp.hpp:19]: also missing *_ suffix

File: src/util/encoding_rlp.hpp
Comment (MatusKysel) [inline src/util/encoding_rlp.hpp:121]: remove & from here also up in few more loop like this. This will trigger warning under clang  (https://github.com/Taraxa-project/taraxa-node/pull/794)

File: src/storage/db_storage.hpp
Comment (02p01r) [reply src/storage/db_storage.hpp:73]: I believe the "_" suffix is only for private fields. Perhaps I got this convention wrong though. To me, a public field with underscore seems quite weird when you access it from outside.

File: src/util/encoding_rlp.hpp
Comment (02p01r) [reply src/util/encoding_rlp.hpp:19]: https://github.com/Taraxa-project/taraxa-node/pull/843#discussion_r648598852

File: src/util/encoding_rlp.hpp
Comment (02p01r) [reply src/util/encoding_rlp.hpp:121]: ok

File: src/CMakeLists.txt
Comment (JakubFornadel) [reply src/CMakeLists.txt:55]: yes there is. To see headers for IDE-generated projects you have to provide list of headers. Just put it back
https://stackoverflow.com/questions/36174499/why-add-header-files-into-add-library-add-executable-command-in-cmake

File: src/network/rpc/EthClient.h
Comment (JakubFornadel) [reply src/network/rpc/EthClient.h:11]: so there is no possibility to set 2 space padding in tool that generated it ?

File: src/common/types.hpp
Comment (JakubFornadel) [reply src/common/types.hpp:28]: I think it should not be in taraxa namespace at all but I see that this practice is broadly used almost everywhere so it might be more of a task for bigger refactor...

File: tests/pbft_manager_test.cpp
Comment (JakubFornadel) [reply tests/pbft_manager_test.cpp:67]: Really ? Because when I joined the project I was told that variables are underscore, class names and functions are camel-case

File: src/network/rpc/eth/impl/Watches.hpp
Comment (JakubFornadel) [reply src/network/rpc/eth/impl/Watches.hpp:21]: This is more about conventions. I will quote Stroustrup: "structs are primarily used for data structures where the members can take any value; that is, we can't define any meaningful invariant "  - you can find this in book "Principles and Practice".

In other words struct in modern C++ are used mostly for short data structures where all members are public, someone uses it purely for POD's, etc....

Software engineering is also about making life easier to other devs so if there is sort of 'rule' in cpp community and people are used to it, lets stick to that.

File: src/final_chain/final_chain.cpp
Comment (JakubFornadel) [reply src/final_chain/final_chain.cpp:437]: ok for factory function

File: src/storage/db_storage.hpp
Comment (JakubFornadel) [reply src/storage/db_storage.hpp:73]: I believe you got it wrong. Underscore is used to easily distinguish between class and non-class variables. All class members should have _ suffix. 

https://google.github.io/styleguide/cppguide.html#Variable_Names - "Data members of classes (but not structs) additionally have trailing underscores. For instance: a_local_variable, a_struct_data_member, a_class_data_member_."

File: src/node/full_node.hpp
Comment (JakubFornadel) [reply src/node/full_node.hpp:98]: Ok thanks for explanation. I get the reasoning although I do think that this mechanism lost meaning once you fixed problems with shared pointers... But that will be separate task to refactor this probably...

File: tests/CMakeLists.txt
Comment (JakubFornadel) [reply tests/CMakeLists.txt:9]: well plain TODO does not say much to me of what should be eventually done. But if I understand you correctly you are going to finish these python tests and uncomment it before its merged so ok :) Will keep this convo opened then...

File: src/node/full_node.cpp
Comment (JakubFornadel) [reply src/node/full_node.cpp:244]: yeah this whole close() function might be in destructor I guess without the need to manually reset shared ptrs, although the author maybe implemented it this way for some reason...

File: src/util/encoding_rlp.hpp
Comment (02p01r) [reply src/util/encoding_rlp.hpp:19]: Is this a misuse of `struct`? If not, then the naming is correct as per the Google naming convention. @JakubFornadel @MatusKysel please confirm

Comment (MatusKysel) [reply src/util/encoding_rlp.hpp:19]: I do no think that this is misuse, so this should be alright as it is

File: src/CMakeLists.txt
Comment (02p01r) [reply src/CMakeLists.txt:55]: done

File: src/network/rpc/EthClient.h
Comment (02p01r) [reply src/network/rpc/EthClient.h]: done - included `.h` files for clang-format

File: src/network/rpc/NetClient.h
Comment (02p01r) [reply src/network/rpc/NetClient.h]: https://github.com/Taraxa-project/taraxa-node/pull/843#discussion_r649886637

File: src/network/rpc/TestClient.h
Comment (02p01r) [reply src/network/rpc/TestClient.h:7]: @JakubFornadel please resolve - can't influence the generator

File: tests/pbft_manager_test.cpp
Comment (02p01r) [reply tests/pbft_manager_test.cpp:67]: Agreed offline that we don't aim at function naming consistency at this point, please resolve @JakubFornadel

File: src/node/full_node.cpp
Comment (02p01r) [reply src/node/full_node.cpp:244]: Improved the comment, please resolve

File: src/network/rpc/eth/impl/Watches.hpp
Comment (02p01r) [reply src/network/rpc/eth/impl/Watches.hpp]: done

File: src/network/rpc/CMakeLists.txt
Comment (02p01r) [reply src/network/rpc/CMakeLists.txt]: done

File: src/final_chain/final_chain.cpp
Comment (02p01r) [reply src/final_chain/final_chain.cpp]: done

File: src/final_chain/final_chain.cpp
Comment (02p01r) [reply src/final_chain/final_chain.cpp]: done

File: src/final_chain/final_chain.cpp
Comment (02p01r) [reply src/final_chain/final_chain.cpp]: done

File: src/final_chain/final_chain.cpp
Comment (02p01r) [reply src/final_chain/final_chain.cpp]: done with the `new`

File: src/util/encoding_rlp.hpp
Comment (02p01r) [reply src/util/encoding_rlp.hpp]: done

**Comment by 02p01r** [comment]
@Taraxa-project/ledger Also went over all known struct/class field naming problems that were not mentioned and fixed them

File: src/storage/db_storage.hpp
Comment (02p01r) [reply src/storage/db_storage.hpp]: done

**Comment by 02p01r** [comment]
@Taraxa-project/ledger Fixed the test performance degradation.

**Comment by 02p01r** [comment]
> @Taraxa-project/ledger Fixed the test performance degradation.

The explanation of the perf issue:
When I first worked on this PR ~4 months ago, I added an uninterruptible sleep in one place not realizing what I did, and forgot about it.
It was in Watches.hpp (then renamed to watches.hpp/cpp). The watches can be installed via json-rpc API, and they have to have an idle timeout. When the idle timeout is hit, we have to remove those stale watches so to be nice about the resources. So I made it in a daemon thread that periodically does the cleanup. Obviously it cannot busy-spin all the time, so I added a sleeping interval, but that sleep was uninterruptible (`std::this_thread::sleep_for` - OS-level sleep).
I changed it to a sleep through `std::condition_variable` (wait with timeout). When the destructor of Watches is called, I notify that condition variable, so the destructor exists ASAP (it joins the watch cleaner thread). Now it's okay

**Comment by 02p01r** [comment]
@Taraxa-project/ledger Moved the python tests to another PR https://github.com/Taraxa-project/taraxa-node/pull/861

**Comment by 02p01r** [comment]
> > Seems there might be actually a lot of good stuff here, just the way this PR is done degrades the overall quality a lot... I dont even know where and how to do some logical review here as there is so many lines and so much stuff mixed here... I tend to agree with Matus that this should have been split into multiple different PR's with proper commits descriptions etc...
> 
> As discussed offline, I will simulate a history of well-defined and build-able commits based on these changes

I attempted to split it into several independent build-able commits, but I cannot achieve this at all:
All the changes in taraxa_node have to stem from the taraxa_aleth changes, where I removed all the high-level ethereum code, including ethereum json-rpc impl.
That removed stuff in taraxa_aleth has to have an immediate answer in taraxa_node. The replacement for it is the new src/final_chain package and the new ethereum json-rpc implementation code that depends on final_chain. src/final_chain  also depends on the src/util package changes (encoding_rlp.hpp, event.hpp). It also depends on taraxa Transaction and DbStorage changes.
PbftManager changes immediately depend on final chain, because FinalChain interface has changed.
The rest of code changes just stem from the above (renamings, removing some includes, etc.)

**Review by MatusKysel — APPROVED** [review]

**Review by JakubFornadel — APPROVED** [review]

**Review by JakubFornadel — CHANGES_REQUESTED** [review]
@02p01r please squash before merging. Squash & merge button is disabled

**Comment by 02p01r** [comment]
> @02p01r please squash before merging. Squash & merge button is disabled

Done, please check.

**Review by JakubFornadel — APPROVED** [review]

**Pull request merged**
